pravega/pravega
# Issue 6598: Avoid throttling system critical operations

**Change log description**  
This PR fixes in vast majority of cases; situations where we see system critical operations get throttled by the throttler in SegmentStore.

**Purpose of the change**  
Fixes Issue #6598 

**What the code does**  
In the doRun method of  OperationProcessor.java is where we start a constant loop of : 
How long to throttle if required --> take items off the OperationQueue --> process these Operations.

Specifically in the throttle part of the above loop, we check if the OperationQueue has any operations that are exempt from 
throttling. If yes we suspend the throttler and proceed with taking these operations off the OperationQueue. Seems like while doing this, there is a possibility where we can get a throttling exempt operation AFTER we have made the decision to not suspend(no throttling exempt operations found at the peek of operationQueue) and BEFORE we start the throttler with some delay, in which cases we end up throttling the system critical operation.

The fix basically tries avoiding this by using the API to suspsend the throttler(notifyThrottlerSourceChanged), just before starting the delayed future which induces the delay.

PS:- Note that not every system critical operation gets throttled. Based on a test done on a local VM where the above throttling usecase was reproduced for a limited time, one could see some 4-5 system critical operatiosn throttled out of 32k total system critical operations for example.

**How to verify it**  
Execute the steps in #6598 with the change. We should not see any system critical operations being throttled provided we have a means to check either using logs or metrics.

Metrics from a local run:-

We know that most of the delay when throttled comes from operations waiting in the operationQueue. This can be seen in the Grafana metric pertaining to OperationQueue wait times. 
This metric on the local system where this PR was tested was modified to report wait times for System Critical operations only.
Here is a metric graph for a sample run in case of both before and after the change.
Before Change:-
![image](https://user-images.githubusercontent.com/88840194/158350639-a259eee5-ecb3-4f87-ab9c-355a3aacc3b0.png)

After Change:-
![image](https://user-images.githubusercontent.com/88840194/158350737-f77321dc-1ef5-413b-9962-4ebd8bdf9cc1.png)


As we can see before the change the operationqueue wait times have been in the order of 10's of seconds. There have also been latencies of the order of 25s as well, but not captured in the graph above; but after the change the operationqueue wait times were mostly under a second.
